### PR TITLE
Update buildsystem to provide out-of-the-box Mono support

### PR DESCRIPTION
**Pull request by whitequark**
After these (fairly minor) changes, all demo code compiles and works in all four mode (even Winforms, even though it looks hideous). See individual commits for details.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -81,11 +81,8 @@ Requirements
     with the version 3.5 of the framework, but CSML was initially
     developped against version 2.0.
 
-  - Alternatively, CSML should work under Linux with Mono and
-    hopefully the newest .NET Core framework (again probably with
-    minor adjustments to the build system).  Support for Mono was only
-    tested in the early days in the project and never used in
-    production, though.
+  - Alternatively, CSML has been under Linux with Mono 4.x.
```
Comment (alainfrisch) [inline README.md]: "has been _tested_"?

Also: any reason to drop the note about the .NET Core framework?

File: Makefile.common [hunk 2]
```diff
@@ -21,9 +21,10 @@ ifeq ($(SYSTEM), win32)
...
   CSMLCSCODE = $(shell cygpath -w $(CSMLDIR)/csml.cs)
   EXECPREFIX=
-
+  PREFIX_DLL=
 else
-  CSC=gmcs
+  CSC=mcs
   CSMLCSCODE = $(CSMLDIR)/csml.cs
   EXECPREFIX=LD_LIBRARY_PATH=.
+  PREFIX_DLL=lib
```
Comment (alainfrisch) [inline Makefile.common]: Can you explain the rationale for the PREFIX_DLL under Unix?  Is it only to follow some convention, or does it address some issue?  Any reason not to do the same under Windows (if only to simplify the doc)?

File: .gitignore [hunk 3]
```diff
@@ -9,4 +9,23 @@
 *.cmx
 *.cmxs
 *.cmxa
+*.cds
 *.exe
+*.so
+*.dll
+/src/csml
```
Comment (alainfrisch) [inline .gitignore:16]: I guess one could add /src/csml.exe.

**Comment by alainfrisch** [comment]
Thanks a lot!  This all look good to me.  Just waiting for your comments on the small notes I've added.

File: README.md (hunk 1 above)
Comment (whitequark) [reply README.md]: Yes, I accidentally a word. Fixed now.

Regarding .NET Core framework: as far as I understand, Mono doesn't need Core as it has its own implementations of .NET fundamentals, and I'm not sure if there's any reason to use Microsoft's version of those (or even if they will work). If I misunderstood it, please correct me and I'll try to rephrase it better.

File: Makefile.common (hunk 2 above)
Comment (whitequark) [reply Makefile.common]: As far as I understand, the reason for this is as follows. On Windows, a library named "m" will generally be in a file "m.dll". On Linux, it would be in a file "libm.so" (or a variation thereof). So, when you write `[DllImport("m")]`, .NET on Windows will transform that into "m.dll", and Mono on Linux will transform that into "libm.so". In case of CSML, an explicit extension is applied, and Mono does not add a second ".so", but it still adds "lib".

The precise set of rules for mangling library names is long and very messy; it is described in detail on this [Mono documentation page](http://www.mono-project.com/docs/advanced/pinvoke/). I'll quote the gist of it:

> How do you write portable Platform Invoke (P/Invoke) code that will work cross-platform?
> The short answer is that you don’t.

It appears that, indeed, `[DllImport("libfoo")]` would do the right thing on both Windows (append ".dll" as long as "foo" doesn't contain any dots) and Linux (append ".so" and don't prepend "lib"), so I could try doing that.

I don't have a Windows OCaml development environment on hand (I cross-compile everything with my [opam-cross-windows](https://github.com/whitequark/opam-cross-windows) repository), so could you test such a change for me once I do it?

File: .gitignore (hunk 3 above)
Comment (whitequark) [reply .gitignore:16]: There's a `*.exe` rule above... But sure, I'll add it.

File: README.md (hunk 1 above)
Comment (alainfrisch) [reply README.md]: AFAIU (and I'm really an expert of this ecosystem), I think these are two different things: Mono implements the .NET Framework, and .NET Core is a new framework (which happens to be very close from .NET) re-designed to be cross-platform.  So I'd rephrase as:

> Alternatively, CSML has been tested under Linux with Mono 4.x, but not used in production.
> 
> Moreover, it is expected that CSML could be made to work with little effort with the newest .NET Core framework both under Windows and other systems.

Comment (whitequark) [reply README.md]: Ah, I understand now. I've updated this to use your phrasing.

File: Makefile.common (hunk 2 above)
Comment (alainfrisch) [reply Makefile.common]: > so could you test such a change for me once I do it?

Yes, of course!

File: .gitignore (hunk 3 above)
Comment (alainfrisch) [reply .gitignore:16]: Sorry, I missed it.

**Comment by whitequark** [comment]
I believe I've addressed all concerns.

**Comment by alainfrisch** [comment]
Just to be sure that I understand:  Mono does not add an extra "lib" prefix only because the specified name already has this prefix?

**Comment by whitequark** [comment]
As far as I understand, yes. I can probably dig up the exact code it uses to mangle the library name if you want.

**Comment by alainfrisch** [comment]
This is quite nasty... I'm not sure I want to force a "lib" prefix (which does not correspond to how we use CSML today, i.e. our ".dll" does not start with "lib").  So, two questions:
- Would Mono also add the "lib" prefix automatically even if the ".so" suffix is explicit in the DllImport attribute?  If no, one could simply pass this suffix in build instructions.
- If there is no way with Mono to have DllImport refer to a .so file without a "lib" prefix, I would suggest to use this "lib" prefix for the standard .so/.dll files, but not add the "lib" prefix automatically in CSML (and document the behavior so that people that require Mono support explicitly use a "lib" prefix).

**Comment by whitequark** [comment]
> Would Mono also add the "lib" prefix automatically even if the ".so" suffix is explicit in the DllImport attribute? If no, one could simply pass this suffix in build instructions.

It would. Horrifyingly enough it doesn't even care to append it to a basename, `[DllImport("a/b")]` gets transformed into `liba/b.so`...

> If there is no way with Mono to have DllImport refer to a .so file without a "lib" prefix, I would suggest to use this "lib" prefix for the standard .so/.dll files, but not add the "lib" prefix automatically in CSML (and document the behavior so that people that require Mono support explicitly use a "lib" prefix).

I believe there is no way. I will perform this change.

**Comment by whitequark** [comment]
I've implemented your suggestion.

**Comment by alainfrisch** [comment]
Thanks again!
